pints/_mcmc/_nuts.py

```diff
@@ -0,0 +1,753 @@
+        # encounter a U-Turn, or reach a max number of points 2^max_tree_depth
+        while j < max_tree_depth and state.s == 1:
+
+            # pick a random direction to integrate in
+            # (to maintain detailed balance)
+            if np.random.randint(0, 2):
+                vj = 1
+            else:
+                vj = -1
+
+            # recursivly build up tree in that direction
+            state_dash = yield from \
+                build_tree(state, vj, j, adaptor,
+                           hamiltonian0, hamiltonian_threshold)
+
+            state.update(state_dash, direction=vj, root=True)
+
+            j += 1
+
+        # update current position in chain
+        theta = state.theta
+        L = state.L
+        grad_L = state.grad_L
+
+        # adapt epsilon and mass matrix using dual averaging
+        restart_stepsize_adapt = \
+            adaptor.step(state.theta, state.alpha / state.n_alpha)
+        if restart_stepsize_adapt:
+            epsilon = yield from \
+                find_reasonable_epsilon(theta, L, grad_L,
+                                        adaptor.get_inv_mass_matrix())
+            adaptor.init_adapt_epsilon(epsilon)
+
+        # signal calling process that mcmc step is complete by passing a tuple
+        # (rather than an ndarray)
+        yield (theta,
+               L,
+               state.alpha / state.n_alpha,
+               state.n_alpha,
+               state.divergent)
+
+        # next step
+        m += 1
+
+
+class NoUTurnMCMC(pints.SingleChainMCMC):
+    r"""
+
+    Implements No U-Turn Sampler (NUTS) with dual averaging, as described in
+    Algorithm 6 in [1]_.
+
+    Implements the multinomial sampling suggested in [2]_. Implements a mass
+    matrix for the dynamics, which is detailed in [2]_. Both the step size and
+    the mass matrix is adapted using a combination of the dual averaging
+    detailed in [1]_, and the windowed adaption for the mass matrix and step
+    size implemented in the Stan library (https://github.com/stan-dev/stan).
+
+    Like Hamiltonian Monte Carlo, NUTS imagines a particle moving over negative
+    log-posterior (NLP) space to generate proposals. Naturally, the particle
+    tends to move to locations of low NLP -- meaning high posterior density.
+    Unlike HMC, NUTS allows the number of steps taken through parameter space
+    to depend on position, allowing local adaptation.
+
+    Note: This sampler is only supported on Python versions 3.3 and newer.
+
+    Extends :class:`SingleChainMCMC`.
+
+    References
+    ----------
+    .. [1] Hoffman, M. D., & Gelman, A. (2014). The No-U-Turn sampler:
+           adaptively setting path lengths in Hamiltonian Monte Carlo.
+           Journal of Machine Learning Research, 15(1), 1593-1623.
+
+    .. [2] Betancourt, M. (2018). `A Conceptual Introduction to Hamiltonian
+           Monte Carlo`, https://arxiv.org/abs/1701.02434.
+
+    """
+
+    def __init__(self, x0, sigma0=None):
```

Comment:
I'm going to assume this is quite similar to when I reviewed ages ago, and that Ben checked the maths :sweat_smile: